spans_buffer.go

```diff
@@ -1,57 +1,74 @@
 package lightstep
 
-import "github.com/opentracing/basictracer-go"
+import (
+	"time"
 
-const defaultMaxSpans = 1000
+	"github.com/opentracing/basictracer-go"
+)
 
 type spansBuffer struct {
-	rawSpans      []basictracer.RawSpan
-	maxBufferSize int
+	rawSpans       []basictracer.RawSpan
+	dropped        int64
+	reportOldest   time.Time
+	reportYoungest time.Time
 }
 
-func (b *spansBuffer) setDefaults() {
-	b.maxBufferSize = defaultMaxSpans
-	b.rawSpans = make([]basictracer.RawSpan, 0, b.maxBufferSize)
+func newSpansBuffer(size int) (b spansBuffer) {
+	b.rawSpans = make([]basictracer.RawSpan, 0, size)
+	b.reportOldest = time.Time{}
+	b.reportYoungest = time.Time{}
+	return
 }
 
-func (b *spansBuffer) setMaxBufferSize(size int) {
-	b.maxBufferSize = size
+func (b *spansBuffer) isHalfFull() bool {
+	return len(b.rawSpans) > cap(b.rawSpans)/2
 }
 
-func (b *spansBuffer) len() int {
-	return len(b.rawSpans)
+func (b *spansBuffer) setCurrent(now time.Time) {
+	b.reportOldest = now
+	b.reportYoungest = now
 }
 
-func (b *spansBuffer) cap() int {
-	return b.maxBufferSize
+func (b *spansBuffer) setFlushing(now time.Time) {
+	b.reportYoungest = now
 }
 
-func (b *spansBuffer) reset() {
-	// Reuse the existing buffer if it's the correct size
-	if cap(b.rawSpans) == b.maxBufferSize {
-		b.rawSpans = b.rawSpans[:0]
-	} else {
-		b.rawSpans = make([]basictracer.RawSpan, 0, b.maxBufferSize)
-	}
+func (b *spansBuffer) clear() {
+	b.rawSpans = b.rawSpans[:0]
+	b.reportOldest = time.Time{}
+	b.reportYoungest = time.Time{}
+	b.dropped = 0
 }
 
-func (b *spansBuffer) current() []basictracer.RawSpan {
-	dst := make([]basictracer.RawSpan, len(b.rawSpans))
-	copy(dst, b.rawSpans)
-	return dst
+func (b *spansBuffer) addSpan(span basictracer.RawSpan) {
+	if len(b.rawSpans) == cap(b.rawSpans) {
+		b.dropped++
+		return
+	}
+	b.rawSpans = append(b.rawSpans, span)
 }
 
-// addSpans returns the number of spans dropped (0 if all were added to the
-// buffer).
-func (b *spansBuffer) addSpans(spans []basictracer.RawSpan) (droppedSpans int) {
-	space := b.maxBufferSize - len(b.rawSpans)
-	count := space
-	if len(spans) < count {
-		count = len(spans)
+func (b *spansBuffer) mergeUnreported(a *spansBuffer) {
```

Comment:
Really a pet peeve and not an actual issue: It took me an extra minute to comprehend what was going on since I have this idea that `a` should be the thing `b` is being merged into (if that makes sense).